### PR TITLE
Nation validatation for nation Name and Url (connects #191)

### PR DESCRIPTION
**Pull request by razu9861**

### CODE REVIEW

File: design/nations/nation-validators.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+unction (newDoc, oldDoc, userCtx, secObj) {
```
Comment (lmmrssa) [inline design/nations/nation-validators.js]: it should be function 'typo error'

File: design/nations/nation-validators.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  function require(field) {
+    var errMessage = field + ' is required';
+    if (!newDoc[field]) {
+      throw { forbidden: errMessage };
+    }
+  }
+  function isHex(field) {
```
Comment (lmmrssa) [inline design/nations/nation-validators.js]: why is this validation of hex here?

File: src/app/nation/nation.component.ts [hunk 3]
```diff
@@ -42,20 +42,22 @@ export class NationComponent implements OnInit {
 
   createForm() {
     this.nationForm = this.fb.group({
-      adminName: [ '', Validators.required,
-        // an arrow function is for lexically binding 'this' otherwise 'this' would be undefined
-        ac => this.nationValidatorService.nationCheckerService$(ac)
+      adminName: [ '', Validators.required
```
Comment (lmmrssa) [inline src/app/nation/nation.component.ts]: it should be on same line or above should be on different lines

File: src/app/validators/nation-validator.service.ts [hunk 4]
```diff
@@ -15,9 +15,9 @@ export class NationValidatorService {
   readonly dbName = 'nations';
 
   constructor(private couchService: CouchService) {}
-    public nationCheckerService$(name: string): Observable<boolean> {
+    public nationCheckerService$(name: string, nationUrl: string): Observable<boolean> {
```
Comment (lmmrssa) [inline src/app/validators/nation-validator.service.ts]: i don't see this function used anymore.

File: src/app/validators/nation-validator.service.ts [hunk 5]
```diff
@@ -15,9 +15,9 @@ export class NationValidatorService {
...
       const isDuplicate = this.couchService
-      .post(`${this.dbName}/_find`, findOneDocument('name', name))
+      .post(`${this.dbName}/_find`, findAllDocuments('name', name, 'nationUrl', nationUrl))
```
Comment (paulbert) [inline src/app/validators/nation-validator.service.ts]: This will not work, since `findAllDocuments` only takes two arguments: one `selector` and one `query`.

File: src/app/validators/nation-validator.service.ts (hunk 4 above)
Comment (paulbert) [inline src/app/validators/nation-validator.service.ts]: `nationCheckerService` is only passed one `AbstractControl` at a time, so it will not receive both the `name` and `nationUrl`.

File: src/app/nation/nation.component.ts [hunk 6]
```diff
@@ -124,13 +119,8 @@ export class NationComponent implements OnInit, AfterViewInit {
           formdata[ '_id' ] = data.id;
           formdata[ '_rev' ] = data.rev;
           this.nations.data.push(formdata);
-          console.log(this.nations.data);
+          this.nations.data = this.nations.data;
```
Comment (lmmrssa) [inline src/app/nation/nation.component.ts]: i think it was changed with some other event in one of comment. So why is this here again?

File: src/app/nation/nation.component.ts [hunk 7]
```diff
@@ -146,16 +136,16 @@ export class NationComponent implements OnInit, AfterViewInit {
     const validation = {
       adminName: [ '', Validators.required,
         // an arrow function is for lexically binding 'this' otherwise 'this' would be undefined
-        ac => this.nationValidatorService.nationCheckerService$(ac)
       ],
-      name: [ '', Validators.required ],
-      nationUrl: [ '', Validators.required ],
+      name: [ '', Validators.required, ac => this.nationValidatorService.checkNationExists$(ac) ],
```
Comment (lmmrssa) [inline src/app/nation/nation.component.ts]: it was supposed to use mango queries selector. but it is still using single field query

File: src/app/nation/nation.component.ts [hunk 8]
```diff
@@ -63,23 +64,18 @@ export class NationComponent implements OnInit, AfterViewInit {
...
-      name: [ '', Validators.required ],
-      nationUrl: [ '', Validators.required ],
-    });
-  }
-
   getNationList() {
     this.couchService.get(this.dbName + '/_all_docs?include_docs=true')
       .then((data) => {
+        const datas = [];
+        for (let i = 0; i < data.rows.length; i++) {
```
Comment (lmmrssa) [inline src/app/nation/nation.component.ts]: if condition and map can just be joined together to form one map function

File: src/app/nation/nation.component.ts (hunk 6 above)
Comment (Rupesh87) [reply src/app/nation/nation.component.ts]: Raju : i kept this code in my pr just to test  sorry i forget to remover it before creating PR

File: design/nations/nation-validators.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+  "validate_doc_update":
+    "function (newDoc, oldDoc, userCtx, secObj) {\n  function require(field) {\n   var errMessage = field + ' is required';\n    if (!newDoc[field]) {\n      throw { forbidden: errMessage };\n    }\n  }\n  function isHex(field) {\n    var val = newDoc[field];\n    var errMessage = field + ' is not a valid hex';\n\n    if (!/^#[A-F0-9]{6}$/i.test(val)) {\n     throw { forbidden: errMessage };\n    }\n  }\n    if (!newDoc) {\n      require('name');\n      require('adminName');\n      require('nationUrl');\n   }\n}"
```
Comment (lmmrssa) [inline design/nations/nation-validators.json]: this file still has hex function

File: src/app/nation/nation.component.ts [hunk 10]
```diff
@@ -137,15 +141,16 @@ export class NationComponent implements OnInit, AfterViewInit {
     const validation = {
       adminName: [ '', Validators.required,
         // an arrow function is for lexically binding 'this' otherwise 'this' would be undefined
```
Comment (lmmrssa) [inline src/app/nation/nation.component.ts]: this comment is not supposed to be here

File: src/app/shared/dialogs/dialogs-form.component.ts [hunk 11]
```diff
@@ -23,7 +23,22 @@ export class DialogsFormComponent {
   public validation: any;
   public message: string;
   public modalForm: FormGroup;
+  readonly dbName = 'nations';
```
Comment (lmmrssa) [inline src/app/shared/dialogs/dialogs-form.component.ts]: this is not supposed to be readonly field as dialog-form could work with any other modal

File: src/app/nation/nation.component.ts [hunk 12]
```diff
@@ -71,7 +73,12 @@ export class NationComponent implements OnInit, AfterViewInit {
       .then((data) => {
         // _all_docs returns object with rows array of objects with 'doc' property that has an object with the data.
         // Map over data.rows to remove the 'doc' property layer
-        this.nations.data = data.rows.map(nation => nation.doc);
+        this.nations.data = data.rows.reduce((nations: any[], nation: any) => {
```
Comment (paulbert) [inline src/app/nation/nation.component.ts]: My mistake earlier, this is not equivalent to only map, but I think using map & filter will be faster & simpler than reduce:

https://jsperf.com/map-filter-vs-reduce-simple
